### ISSUE TITLE
Weird behavior in files with top-level statements

### ISSUE THREAD

**Issue by bjornhellander** [DotNetAnalyzers/StyleCopAnalyzers#4068]
Created this from PR comment https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/4054#discussion_r2590409994.

Consider a console app with this code in Program.cs:

```
// <copyright file="Program.cs" company="PlaceholderCompany">
// Copyright (c) PlaceholderCompany. All rights reserved.
// </copyright>

/// <summary>
/// Program.
/// </summary>
internal static class Program
{
    /// <summary>
    /// Program entry point.
    /// </summary>
    public static void Main()
    {
        var x = new MyData();
        //// Do something more here
    }

    private class MyData
    {
    }
}
```

This triggers no StyleCop diagnostics.

If I simplify  this to use top-level statements, I get this:

```
// <copyright file="Program.cs" company="PlaceholderCompany">
// Copyright (c) PlaceholderCompany. All rights reserved.
// </copyright>

var x = new MyData();
//// Do something more here

/// <summary>
/// Contains some data.
/// </summary>
internal class MyData
{
}
```

Note that the compiler forces me to change MyData to no longer be private.

This triggers SA1649 (File name should match first type name) which wants the file name to be MyData.cs. Program.cs is still the natural name in my opinion, since the primary responsibility for this file is to define the entry point code. The desired action would in my opinion be to move MyData to a separate file.

I suggest that SA1649 should be changed to allow any file name when containing top-level statements and also to update SA1642 (File may only contain a single type) to also trigger on the first type in files with top-level statements. Otherwise, the diagnostic and available code fix might push users to rename the file instead, which I think is the wrong thing to do.

**Comment by bjornhellander** [comment]
All Microsoft templates that I have checked uses Program.cs. Enforcing that seems fine for me for normal projects.

For single-file applications (https://community.visualstudiomagazine.com/t/single-file-c-applications-are-here/192), this feels undersirable to me, but StyleCop is probably not widely used in those circumstances. (Running analyzers in this scenario might not even be possible, I haven't checked this enough to know.)

**Comment by gtbuchanan** [comment]
I can confirm that analyzers run in single-file apps. I found out on accident when migrating from [dotnet-script](https://github.com/dotnet-script/dotnet-script). We have a repo-level `Directory.Build.props` that applies all of our analyzers, and it started applying them to the single-file app 🙃
